components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.kt

```diff
@@ -0,0 +1,307 @@
+     */
+    class Builder(
+        var id: Id,
+        var originStatsFactory: OriginStatsFactory,
+        var loadBalancer: LoadBalancer,
+        var metrics: CentralisedMetrics,
+        var rewriteRules: List<RewriteRule> = emptyList(),
+        var originsRestrictionCookieName: String? = null,
+        var stickySessionConfig: StickySessionConfig = StickySessionConfig.stickySessionDisabled(),
+        var originIdHeader: CharSequence = StyxHeaderConfig.ORIGIN_ID_DEFAULT,
+        var retryPolicy: RetryPolicy = RetryNTimes(3),
+        var overrideHostHeader: Boolean = false,
+    ) {
+        fun build() = StyxBackendServiceClient(this)
+    }
+
+
+    private val rewriteRuleset: RewriteRuleset = RewriteRuleset(rewriteRules)
+
+    override fun sendRequest(request: LiveHttpRequest, context: HttpInterceptor.Context): Publisher<LiveHttpResponse> =
+        sendRequest(rewriteUrl(request), emptyList(), 0, context)
+
+
+    private fun isError(status: HttpResponseStatus): Boolean = status.code() >= 400
+
+
+    private fun bodyNeedsToBeRemoved(request: LiveHttpRequest, response: LiveHttpResponse): Boolean =
+        isHeadRequest(request) || isBodilessResponse(response)
+
+
+    private fun responseWithoutBody(response: LiveHttpResponse): LiveHttpResponse = response.newBuilder()
+        .header(HttpHeaderNames.CONTENT_LENGTH, 0)
+        .removeHeader(HttpHeaderNames.TRANSFER_ENCODING)
+        .removeBody()
+        .build()
+
+
+    private fun isBodilessResponse(response: LiveHttpResponse): Boolean =
+        response.status().code() == 204 || response.status().code() == 304 || response.status().code() / 100 == 1
+
+
+    private fun isHeadRequest(request: LiveHttpRequest): Boolean = request.method() == HttpMethod.HEAD
+
+    private fun shouldOverrideHostHeader(host: RemoteHost, request: LiveHttpRequest): LiveHttpRequest =
+        if (overrideHostHeader && !host.origin().host().isNullOrBlank()) {
+            request.newBuilder().header(HttpHeaderNames.HOST, host.origin().host()).build()
+        } else request
+
+
+    private fun sendRequest(
+        request: LiveHttpRequest,
+        previousOrigins: List<RemoteHost>,
+        attempt: Int,
+        context: HttpInterceptor.Context
+    ): Publisher<LiveHttpResponse> {
+        if (attempt >= MAX_RETRY_ATTEMPTS) {
+            return Flux.error(NoAvailableHostsException(id))
+        }
+        val remoteHost = selectOrigin(request)
+        return if (remoteHost.isPresent) {
+            val host = remoteHost.get()
+            val updatedRequest = shouldOverrideHostHeader(host, request)
+            val newPreviousOrigins: MutableList<RemoteHost> = ArrayList(previousOrigins)
+            newPreviousOrigins.add(host)
+            ResponseEventListener.from(
+                host.hostClient().handle(updatedRequest, context)
+                    .map { response: LiveHttpResponse -> addStickySessionIdentifier(response, host.origin()) }
```

Comment:
too verbose, don't need to specify types here. Using named argument or the implicit "it" seems like a pretty subjective thing so i would follow project conventions there.